### PR TITLE
[tune] docker syncer

### LINKED ISSUE #4403: [tune/autoscaler] _LogSyncer cannot rsync with Docker
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 18.04
- **Ray installed from (source or binary)**: binary
- **Ray version**: 0.6.4
- **Python version**: 3.6.7

### Describe the problem
In `_LogSyncer`, `sync_to_worker_if_possible` and `sync_now` use `rsync` to transfer logs between the local node and the worker. This breaks when using Docker, since:
  - If the local node is in Docker, it will typically have the `root` username, and so this is what `get_ssh_user` will return. But we cannot typically login to the worker node as `root`.
  - The `local_dir` on the worker is inside the Docker container, and may not even be visible outside. If it is bound, then it will typically be at a different path.

An unrelated issue: if `self.sync_func` is non-None, it will get executed *before* the `worker_to_local_sync_cmd`, which I think is wrong.

I'd be happy to make a stab at a PR, but I'd appreciate some suggestions on the right way of fixing this, as it's been a while since I've looked at Ray internals. This also feels like a problem that is likely to reoccur with slight variation, e.g. this bug is similar to #4183

Perhaps we can make `autoscaler` provide an abstract `sync` interface that `tune` and other consumers can use. This could make to `rsync` in the standard case, and something more complex in the Docker case (e.g. Docker cp followed by rsync)? `ray.autoscaller.commands.rsync` is already something along these lines -- would this be an appropriate place to modify?

A more hacky solution would be to make `get_ssh_user` return the right value and make the Docker volume-binding line up so that we can just ignore the difference between Docker and non-Docker instances.

### Source code / logs
A MWE for this is hard to provide, but if the above description is insufficient I can try to come up with one.

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

Add support for syncing between docker containers in Ray Tune.

## Related issue number

Replaces and closes #10800
Also, closes https://github.com/ray-project/ray/issues/4403
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/tune/user-guide.rst [hunk 1]
```diff
@@ -448,6 +448,27 @@ By default, syncing occurs every 300 seconds. To change the frequency of syncing
...
+Tune automatically syncs files and checkpoints between different remote
+containers as needed.
+
+To make this work in your Docker cluster, e.g. when you are using the Ray autoscaler
+with docker containers, you will need to pass a
+``DockerSyncer`` to the ``sync_to_driver`` argument of ``tune.SyncConfig``.
+
+.. code-block:: python
+
+    from ray.tune.integration.docker import DockerSyncer
```
Comment (richardliaw) [inline doc/source/tune/user-guide.rst:465]: let's be sure to add the docstrings

File: python/ray/tune/integration/docker.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+    def external_ip(self, node_id):
+        return self.provider.external_ip(node_id)
+
+
+_provider = None
+
+
+class DockerSyncer(NodeSyncer):
+    """DockerSyncer used for synchronization between Docker containers.
```
Comment (richardliaw) [inline python/ray/tune/integration/docker.py:67]: Can you make a note that this only works with the Ray cluster launcher?

File: doc/source/tune/user-guide.rst (hunk 1 above)
Comment (richardliaw) [reply doc/source/tune/user-guide.rst:465]: (integrations.rst I think)

File: python/ray/tune/integration/docker.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+            command_runner = self._create_command_runner(node_id)
+            self._command_runners[node_id] = command_runner
+        return self._command_runners[node_id]
+
+    def sync_up(self, source, target):
+        """Here target is a tuple (target_node, target_dir)"""
+        target_node, target_dir = target
+
+        # Add trailing slashes for rsync
+        source += "/" if not source.endswith("/") else ""
```
Comment (richardliaw) [inline python/ray/tune/integration/docker.py:157]:
```suggestion
        source = os.path.join(source, "")
```

nit: could be shorter :)

File: python/ray/tune/integration/docker.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+from ray.autoscaler.tags import NODE_KIND_HEAD, NODE_KIND_UNMANAGED, \
+    NODE_KIND_WORKER, TAG_RAY_NODE_KIND
+from ray.tune.syncer import NodeSyncer
+from ray.tune.sync_client import SyncClient
+
+
+class _WrappedProvider:
+    def __init__(self, config_path):
+        from ray.autoscaler._private.util import validate_config
+        from ray.autoscaler.node_provider import _get_node_provider
```
Comment (richardliaw) [inline python/ray/tune/integration/docker.py:16]: I think if you rebase onto master this will not work - these are now private methods?

Comment (krfricke) [reply python/ray/tune/integration/docker.py:16]: In the refactoring this got introduced into the `KubernetesSyncer`, too. Main question is what the alternative would be - reimplementing seems like a waste? Are there public alternatives to these? (I can look into that tomorrow)

Comment (richardliaw) [reply python/ray/tune/integration/docker.py:16]: I think it's ok to keep as private method and then just raise a concern that we need them (or ideally, we can abstract out a part of the commandrunner?)

File: python/ray/tune/integration/docker.py (hunk 3 above)
Comment (krfricke) [reply python/ray/tune/integration/docker.py:157]: Always more to learn!

**Comment by krfricke** [comment]
I'll add unittesting for the integration tomorrow (similar to the Kubernetes test).

File: python/ray/tune/tests/test_integration_docker.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
+import unittest
+from typing import Dict, Optional
+
+from ray.autoscaler.tags import NODE_KIND_HEAD, NODE_KIND_WORKER, \
+    TAG_RAY_NODE_KIND
+from ray.tune.integration.docker import DockerSyncer, \
+    DockerSyncClient, _WrappedProvider
+from ray.tune.sync_client import SyncClient
+from ray.tune.syncer import NodeSyncer
```
Comment (richardliaw) [inline python/ray/tune/tests/test_integration_docker.py:9]: Oh dont forget to add this to Bazel (BUILD)

File: python/ray/tune/integration/docker.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+class _WrappedProvider:
+    def __init__(self, config_path: str):
+        from ray.autoscaler._private.util import validate_config
+        from ray.autoscaler.node_provider import _get_node_provider
+
+        with open(config_path) as f:
+            new_config = yaml.safe_load(f.read())
+        validate_config(new_config)
+        self._config = new_config
```
Comment (richardliaw) [inline python/ray/tune/integration/docker.py:22]: FYI @ericl for Tune to sync between docker containers, we leverage some of the autoscaler's private interfaces.

Comment (richardliaw) [reply python/ray/tune/integration/docker.py:22]: Please advise if there's a better alternative

Comment (ericl) [reply python/ray/tune/integration/docker.py:22]: Can you instead add validate_config to the public API (sdk.py)?

File: python/ray/tune/integration/docker.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+import yaml
+from ray import services
+from ray.autoscaler._private.command_runner import DockerCommandRunner
+from ray.autoscaler.tags import NODE_KIND_HEAD, NODE_KIND_UNMANAGED, \
+    NODE_KIND_WORKER, TAG_RAY_NODE_KIND
+from ray.tune.syncer import NodeSyncer
+from ray.tune.sync_client import SyncClient
+
+
+class _WrappedProvider:
```
Comment (ericl) [inline python/ray/tune/integration/docker.py:14]: This is pretty sketchy. Let me think about this more.

**Comment by ericl** [comment]
@richardliaw @krfricke can we have a design document on this? It looks like this is a pretty complex piece of machinery here and it's concerning it is peeking into the internals of the autoscaler. This makes it likely to be broken by future changes.

**Review by ericl — DISMISSED** [review]
@richardliaw @krfricke can we have a design document on this? It looks like this is a pretty complex piece of machinery here and it's concerning it is peeking into the internals of the autoscaler. This makes it likely to be broken by future changes.

File: python/ray/tune/integration/docker.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+        """Here target is a tuple (target_node, target_dir)"""
+        target_node, target_dir = target
+
+        # Add trailing slashes for rsync
+        source = os.path.join(source, "")
+        target_dir = os.path.join(target_dir, "")
+
+        command_runner = self._get_command_runner(target_node)
+        command_runner.run_rsync_up(source, target_dir)
+        return True
```
Comment (richardliaw) [inline python/ray/tune/integration/docker.py:172]: We can actually maybe just use the docker command runner code here:

```
        options = options or {}
        host_source = os.path.join(DOCKER_MOUNT_PREFIX, source.lstrip("/"))
        self.ssh_command_runner.run(
            f"mkdir -p {os.path.dirname(host_source.rstrip('/'))}")
        if source[-1] == "/":
            source += "."
            # Adding a "." means that docker copies the *contents*
            # Without it, docker copies the source *into* the target
        if not options.get("file_mount", False):
            self.ssh_command_runner.run("docker cp {}:{} {}".format(
                self.container_name, self._docker_expand_user(source),
                host_source))
        # replace this below line
        self.ssh_command_runner.run_rsync_down(
            host_source, target, options=None)
```

**Comment by richardliaw** [comment]
Let's use https://github.com/ray-project/ray/pull/11122 as the API for rsync.

File: python/ray/tune/tests/test_integration_docker.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+            "w1": "1.0.0.1",
+            "w2": "1.0.0.2"
+        })
+        self.local_dir = "/tmp/local"
+        self.remote_dir = "/tmp/remote"
+
+        self.mock_command = _MockRsync()
+
+        from ray.tune.integration import docker
+        docker.rsync = self.mock_command
```
Comment (richardliaw) [inline python/ray/tune/tests/test_integration_docker.py:63]: looks good, - generally you'll want to use `mock.patch` instead

**Review by richardliaw — APPROVED** [review]

**Pull request merged**